server/src/main/java/org/elasticsearch/action/support/single/instance/InstanceShardOperationRequest.java

```diff
@@ -130,5 +141,16 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeOptionalString(concreteIndex);
     }
 
+    public void writeThin(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        if (shardId != null && index.equals(shardId.getIndexName())) {
```

Comment:
Same here, the `ShardId` is often known on the coordinating node, we can improve this situation to less often serialize the actual name.